### DISCUSSION TITLE
optimized support for Windows

### DISCUSSION THREAD

**Discussion by nnako** [GothenburgBitFactory/taskwarrior#2659] [Ideas]
I am a Windows user. Not, because I wish to be but because my working environment forces me to. Reading the documentation for getting started with TaskWarrior on Windows, I discovered that there seem to be the same problems as many other tools have. All these tools are at least partly developed using Python and use some `curses`-related packages like `blessings`, `urwid`, `npyscreen`, ...

Now, I was wondering if there was some other experiences regarding providing "better" support for the native `batch terminal` on Windows computers (no need for MSYS, CYGWIN, ...). From my personal experience, there are definitely alternatives to the `curses` ecosystem. These alternatives provide good and stable native terminal access on Windows machines without complex installation procedures. Even with nicely colored outputs and able to use the various unicode characters. Examples of these alternative packages are:

- [`prompt-toolkit`](https://github.com/prompt-toolkit/python-prompt-toolkit)
- [`rich`](https://github.com/willmcgugan/rich)
- ...

So,

- what are the causes for the difficulties regarding support for native Windows consoles in the TaskWarrior project? Is it (as suggested) the usage of `curses`-related packages in Python?
- if so, what would be the obstacles to switch the console support to e.g. `prompt-toolkit` and thus get rid of these problems? Would there be other obstacles for this approach?

**Comment by tbabej** [comment]
Hi @nnako! The core taskwarrior does not depend on any terminal UI library, and should work just fine on Windows either using Cygwin or WSL. Perhaps you're having trouble with community extensions, such as `vit` or `taskwarrior-tui`? :thinking:

**Reply by nnako** [reply]
Hi @tbabej! Thanks for your quick reply. Ok. I see, there does not seem to be a dependency of TaskWarrior on the Python packages mentioned above. So, no way to solve anything by these means, here.

But, as you mentioned, TaskWarrior still needs Cygwin (which is not supported anymore, according to the [QuickSetup](https://taskwarrior.org/download/) documentation) or WSL (which I am not allowed to use within my working environment).

I was thinking about how to use TaskWarrior directly in Windows' "native" terminal application which ships with every Windows installation since the birth of Windows itself. It is the one that gets started if you press `<WIN> + <R>` then type `cmd` and hit `<RETURN>`.

**Reply by tbabej** [reply]
You're correct about Cygwin, but if I recall correctly, this used to work in the past, so I think the declaration of support for WSL over Cygwin was mostly to lessen the burden maintenance-wise and not a result of some breaking change.

Also note that there are some helpful discussions about native windows binary in this issue: https://github.com/GothenburgBitFactory/taskwarrior/issues/2159 and it seems like @xeijin was successful in building a native TW binary using `mingw` :slightly_smiling_face: so there is hope!

**Comment by nnako** [comment]
Hi again, I finally was able to successfully build `TaskWarrior` on a Raspberry Pi (which is the only Linux-based system I could get hold of). And still, I would like to ask what specific dependencies make it so difficult for `TaskWarrior` to be built natively on Windows machines. I couldn't build it using `mingw`, yety even with the help mentioned above.

I would be eager to understand the current technical problems / restrictions and their causes. As soon as I do, maybe I could try to support in finding a solution?
